### PR TITLE
[CMake] Add build option "asan" for address sanitizer builds.

### PR DESCRIPTION
**Pull request by hageboeck**
- Add build option `asan` to build with address sanitizer.
Only works with clang so far.
- Add suppression file for leak sanitizer, suppressing fallout from
llvm and clang memory management.
- Set up compiler flags for statically linking address sanitizer
into all ROOT exectuables.
- A micro library for configuring asan was added, and statically linked
into all ROOT executables.
  Address sanitizer (at least in clang) allows to define
  __asan_default_config and __lsan_default_suppressions to set good
  defaults.
- A shared version of the setup library is available for all cases where
the address sanitizer run time is LD_PRELOADed into non-instrumented
executables.
- LD_PRELOAD with asan runtimes has been set up for all tests that use
python as executable.

### CODE REVIEW

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-fedora30/cxx14 with flags -DCMAKE_CXX_COMPILER=clang\+\+ -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-fedora30/cxx14 with flags -DCMAKE_CXX_COMPILER=clang\\\+\\\+ -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-fedora30/cxx14 with flags -DCMAKE_CXX_COMPILER='clang++' -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -DCMAKE_CXX_COMPILER='clang++' -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1804-clangHEAD with flags -DCMAKE_CXX_COMPILER='clang++' -Dasan=ON

File: build/LSan.supp [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#llvm memory hoggers
```
Comment (pcanal) [inline build/LSan.supp]: We have put the equivalent file for valgrind in $BUILD_DIR/etc (and $ROOTSYS/etc) so that they can be re-used by users running the same type of checks.

File: cmake/modules/RootMacros.cmake [hunk 2]
```diff
@@ -16,6 +16,7 @@ if(WIN32)
   set(runtimedir ${CMAKE_INSTALL_BINDIR})
 elseif(APPLE)
   set(ld_library_path DYLD_LIBRARY_PATH)
+  set(ld_preload DYLD_INSERT_LIBRARIES)
```
Comment (Axel-Naumann) [inline cmake/modules/RootMacros.cmake]: Sure this works on Catalina? https://theevilbit.github.io/posts/dyld_insert_libraries_dylib_injection_in_macos_osx_deep_dive/

File: cmake/modules/RootTestDriver.cmake [hunk 3]
```diff
@@ -72,10 +72,9 @@ endif()
 
 #---Set environment --------------------------------------------------------------------------------
 if(ENV)
-  string(REPLACE "@" "=" _env ${ENV})
```
Comment (Axel-Naumann) [inline cmake/modules/RootTestDriver.cmake]: Why did you need to remove this?

File: cmake/modules/RootMacros.cmake (hunk 2 above)
Comment (hageboeck) [reply cmake/modules/RootMacros.cmake]: Absolutely not sure! So far, I was only able to test on ubuntu with clang. Therefore the PR ...

File: cmake/modules/RootTestDriver.cmake (hunk 3 above)
Comment (hageboeck) [reply cmake/modules/RootTestDriver.cmake]: For environment variables à la `ASAN_OPTIONS=verbose=1`, this wreaks havoc. With the regex parsing just below, this isn't necessary, any more. The regex parsing will only split at the first `=`, and all others can stay intact.

Comment (hageboeck) [reply cmake/modules/RootTestDriver.cmake]: However, since I discovered the undocumented `__asan_default_options` weak symbol magic, this might not be necessary, any more. Unless we decide to use `gcc`, that is, which probably doesn't have the weak symbol magic, which would force us to pass environment variables across everything again.

**Comment by hageboeck** [comment]
@phsft-bot build with flags -DCMAKE_CXX_COMPILER='clang++' -Dasan=ON

File: build/LSan.supp (hunk 1 above)
Comment (hageboeck) [reply build/LSan.supp]: In principle, I agree, but valgrind doesn't support globbing on stack traces. The equivalent suppression file for valgrind would by **much** more verbose. If you know an automatic way to generate one, be my guest, but I won't do this manually.

Comment (pcanal) [reply build/LSan.supp]: I am confused.  valgrind *does* support globbing .. but more important I am not asking for a new valgrind file but for **this** file (as is) to be in ./etc to be using by a user doing on its ROOT-using-library the ASAN analyze that you are introducing here.  (I guess it does depend on whether this file is read at run-time or compile-time, if it is the later then my advise is moot)

**Comment by oshadura** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -DCMAKE_CXX_COMPILER='clang++' -Dasan=ON

**Comment by oshadura** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -Dasan=ON

File: core/sanitizer/SanitizerSetup.cxx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ *   WV, Wouter Verkerke, UC Santa Barbara, verkerke@slac.stanford.edu       *
+ *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
+ *                                                                           *
+ * Copyright (c) 2000-2019, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
```
Comment (eguiraud) [inline core/sanitizer/SanitizerSetup.cxx]: does his banner fit in here?

File: core/sanitizer/SanitizerSetup.cxx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+"leak:clang::LineTableInfo \n"
+"leak:clang::HeaderSearch \n"
+"leak:clang::Diag \n"
+"leak:clang::Preprocessor:: \n"
+"leak:clang::TextDiagnosticPrinter \n"
+"leak:clang-tblgen \n"
+"leak:cling::IncrementalExecutor\n"; //cling macro execution
+}
+
+}
```
Comment (eguiraud) [inline core/sanitizer/SanitizerSetup.cxx]: for posterity, maybe add some indication of what this file is?

File: core/sanitizer/SanitizerSetup.cxx (hunk 4 above)
Comment (hageboeck) [reply core/sanitizer/SanitizerSetup.cxx]: Ha, good spot. The shallow reviewing seems to work!

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -Dasan=ON

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default with flags -Dasan=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo

File: build/LSan.supp (hunk 1 above)
Comment (pcanal) [reply build/LSan.supp]: @hageboeck This is marked as resolved but _I_ am still confused :).   

Is build/LSan.supp used only at build time or is it read at run-time>

If read at run-time, is build/LSan.supp of any use for external project?  Namely, if I external project build their project (and possibly ROOT) with the sanitizer on, should they be using build/LSan.supp ?

Thanks.

Comment (hageboeck) [reply build/LSan.supp]: Ok, got it. The suppression file can be used when an external project tries to use a ROOT with instrumentation. They will have to `LD_PRELOAD` or link the sanitizer runtimes, though. If that happens, all the leaks that are suppressed automatically in ROOT would come back.

**Comment by hageboeck** [comment]
@phsft-bot build just on ROOT-ubuntu1904-clang/default mac1015/cxx17 with flags -Dasan=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo

File: build/LSan.supp (hunk 1 above)
Comment (pcanal) [reply build/LSan.supp:3]: > LSAN_OPTIONS=suppressions=$ROOTSYS/etc/LSan.supp 

For symetry reason, consider using the name ```$ROOTSYS/etc/LSan-root.supp``` (and actually maybe ```lsan-root.supp```

File: core/sanitizer/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,22 @@
+if(asan)
+  # Make a shared config. This is needed for executables like python where we have to LD_PRELOAD
```
Comment (eguiraud) [inline core/sanitizer/CMakeLists.txt:2]: s/config/library?

File: core/sanitizer/SanitizerSetup.cxx [hunk 7]
```diff
@@ -0,0 +1,65 @@
+// Author: Stephan Hageboeck, CERN  2 Mar 2020
+
+/*************************************************************************
+ * Copyright (C) 1995-2016, Rene Brun and Fons Rademakers.               *
```
Comment (eguiraud) [inline core/sanitizer/SanitizerSetup.cxx:4]: 2016 :sweat_smile:

File: cmake/modules/SetUpMacOS.cmake [hunk 8]
```diff
@@ -85,6 +85,16 @@ if (CMAKE_SYSTEM_NAME MATCHES Darwin)
...
      set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -bind_at_load -m64")
+     
+     if(asan)
+       #This should be the right way to do it, but clang 10 seems to have a bug
+       #execute_process(COMMAND ${CMAKE_CXX_COMPILER} --print-file-name=libclang_rt.asan_osx_dynamic.dylib OUTPUT_VARIABLE ASAN_RUNTIME_LIBRARY OUTPUT_STRIP_TRAILING_WHITESPACE)
+       execute_process(COMMAND mdfind -name libclang_rt.asan_osx_dynamic.dylib OUTPUT_VARIABLE ASAN_RUNTIME_LIBRARY OUTPUT_STRIP_TRAILING_WHITESPACE)
+       set(ASAN_EXTRA_CXX_FLAGS -fsanitize=address -fno-omit-frame-pointer -fsanitize-blacklist=${CMAKE_SOURCE_DIR}/build/ASan_blacklist.txt)
+      set(ASAN_EXTRA_SHARED_LINKER_FLAGS "-fsanitize=address -static-libsan")
+      #Declaring the __asan_default_options etc as undefined forces the linker to copy them into the executable
+      set(ASAN_EXTRA_EXE_LINKER_FLAGS "-fsanitize=address -static-libsan -Wl,-u,___asan_default_options -Wl,-u,___lsan_default_options -Wl,-u,___lsan_default_suppressions")
```
Comment (eguiraud) [inline cmake/modules/SetUpMacOS.cmake:96]: indentation is weird

File: cmake/modules/RootMacros.cmake (hunk 2 above)
Comment (hageboeck) [reply cmake/modules/RootMacros.cmake:19]: However, that *would* be the correct variable. If you set up your OS X such that you can LD_PRELOAD, that's what you would have to add.

Comment (oshadura) [reply cmake/modules/RootMacros.cmake:19]: Can we try to check somewhere in our infrastructure?

File: cmake/modules/SetUpMacOS.cmake [hunk 9]
```diff
@@ -85,6 +85,17 @@ if (CMAKE_SYSTEM_NAME MATCHES Darwin)
 
      set(CMAKE_C_LINK_FLAGS "${CMAKE_C_LINK_FLAGS} -bind_at_load -m64")
      set(CMAKE_CXX_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -bind_at_load -m64")
+     
+     if(asan)
+       # See also core/sanitizer/Readme.md for what's happening.
```
Comment (oshadura) [inline cmake/modules/SetUpMacOS.cmake:90]: Usually README name of file is capitalized, can you please rename it everywhere (name of file and in the comments)? Thanks!

File: cmake/modules/SetUpMacOS.cmake [hunk 10]
```diff
@@ -85,6 +85,17 @@ if (CMAKE_SYSTEM_NAME MATCHES Darwin)
...
+
```
Comment (oshadura) [inline cmake/modules/SetUpMacOS.cmake:91]: Empty line

File: cmake/modules/SetUpMacOS.cmake [hunk 11]
```diff
@@ -85,6 +85,17 @@ if (CMAKE_SYSTEM_NAME MATCHES Darwin)
...
+       #This should be the right way to do it, but clang 10 seems to have a bug
+       #execute_process(COMMAND ${CMAKE_CXX_COMPILER} --print-file-name=libclang_rt.asan_osx_dynamic.dylib OUTPUT_VARIABLE ASAN_RUNTIME_LIBRARY OUTPUT_STRIP_TRAILING_WHITESPACE)
```
Comment (oshadura) [inline cmake/modules/SetUpMacOS.cmake:93]: hmmm :( did you open bug report? @Axel-Naumann what could be done here?

File: core/sanitizer/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,23 @@
+if(asan)
```
Comment (oshadura) [inline core/sanitizer/CMakeLists.txt:1]: Can we move if(asan)..endif() one step above? in core/CMakeLists.txt?

File: CMakeLists.txt [hunk 13]
```diff
@@ -158,6 +158,21 @@ if (build_timing)
   #set_property(GLOBAL PROPERTY RULE_LAUNCH_CUSTOM "${CMAKE_COMMAND} -E time")
 endif()
 
+#--- Set up address sanitizer builds ----------------------------------------------------------
+if(asan)
+  if(NOT CMAKE_COMPILER_IS_GNUCXX AND NOT CMAKE_CXX_COMPILER_ID MATCHES Clang)
+    message(WARNING "Address sanitizer builds only work with GCC and Clang")
```
Comment (eguiraud) [inline CMakeLists.txt:164]: is it ok to keep going anyway?

File: README/ReleaseNotes/v622/index.md [hunk 14]
```diff
@@ -144,4 +144,8 @@ Not all features of TGeoPainter are supported - only plain drawing of selected T
 
 ## Build, Configuration and Testing Infrastructure
 
+### Experimental address sanitizer build configuration
+Added a build flag `asan` that switches on address sanitizer. It's experimental, so expect problems (e.g. cannot build with gcc because gcc doesn't allow blacklisting).
```
Comment (eguiraud) [inline README/ReleaseNotes/v622/index.md:148]: can now build on gcc right?

File: CMakeLists.txt (hunk 13 above)
Comment (hageboeck) [reply CMakeLists.txt:164]: Rephrased to "only tested". Maybe it will work, but we don't promise anything.

File: README/ReleaseNotes/v622/index.md (hunk 14 above)
Comment (hageboeck) [reply README/ReleaseNotes/v622/index.md:148]: Clarified

File: cmake/modules/RootMacros.cmake (hunk 2 above)
Comment (hageboeck) [reply cmake/modules/RootMacros.cmake:19]: Do you know how to disable source fortification or the other protection things on Mac? If somebody has this disabled, this variable *should* do the right thing. I thought it's better to provide something that *should* work than not to provide anything at all. Realistically, this won't work in our infrastructure for the foreseeable future, though.

File: cmake/modules/SetUpMacOS.cmake (hunk 10 above)
Comment (hageboeck) [reply cmake/modules/SetUpMacOS.cmake:91]: Is that a problem? Thought it would be nice to separate the info on what's happening from the attempt to do it right, which fails to do the clang bug.

File: cmake/modules/SetUpMacOS.cmake (hunk 11 above)
Comment (hageboeck) [reply cmake/modules/SetUpMacOS.cmake:93]: No bug report. Had enough other things to do. Just wanted to get this working.

File: core/sanitizer/CMakeLists.txt (hunk 12 above)
Comment (hageboeck) [reply core/sanitizer/CMakeLists.txt:1]: Sure.

File: cmake/modules/SetUpMacOS.cmake (hunk 10 above)
Comment (oshadura) [reply cmake/modules/SetUpMacOS.cmake:91]: I am not picky, just noticed, thanks  for explanation :)

**Comment by hageboeck** [comment]
> Looks good. Commit history still needs to be cleaned-up

I thought that this would be obvious. I had actually planned to use github's "Squash on merge", but the "Changes requested" prevents that.

**Comment by eguiraud** [comment]
I thought "squash and merge" combined _all_ commits in the PR into a single one?
